util/dvsim/FpvCfg.py

```diff
@@ -0,0 +1,277 @@
+# Copyright lowRISC contributors.
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+
+import logging as log
+from pathlib import Path
+
+import hjson
+from tabulate import tabulate
+
+from OneShotCfg import OneShotCfg
+from utils import subst_wildcards
+
+
+class FpvCfg(OneShotCfg):
+    """Derivative class for FPV purposes.
+    """
+    def __init__(self, flow_cfg_file, proj_root, args):
+        super().__init__(flow_cfg_file, proj_root, args)
+        self.header = ["name", "fpv_errors", "fpv_warnings", "proven", "cex", "undetermined",
+                       "covered", "unreachable", "pass_rate", "cov_rate"]
+        self.summary_header = ["name", "pass_rate", "stimuli_cov", "coi_cov", "prove_cov"]
+
+    def __post_init__(self):
+        super().__post_init__()
+        self.results_title = self.name.upper() + " FPV Results"
+
+    def parse_dict_to_str(self, input_dict, excl_keys = []):
+        # This is a helper function to parse dictionary items into a string.
+        # This function has an optional input "excl_keys" for user to exclude
+        # printing out certain items according to their keys.
+        # Note this function did not sort the input dictionary's key value
+        # before printing the keys and items. If input dictionary is not an
+        # OrderedDictionary, print out key order is not predictable.
+        # This function works for Hjson lib outputs because the lib uses an
+        # OrderDict when it reads dictionaries.
+        # Example Input:
+        # {
+        #   "unreachable": ["prop1, prop2, prop3"],
+        #   "cex"        : ["prop1"],
+        # }
+        # Example Output:
+        # string = "unreachable:
+        # ```
+        # prop1
+        # prop2
+        # prop3
+        # ```
+        # cex:
+        # ```
+        # prop1
+        # ```"
+        output_str = ""
+        for key, item in input_dict.items():
+            if (key not in excl_keys) and item:
+                output_str += "\n" + key + ":\n"
+                output_str += "```\n"
+                output_str += "\n".join(item)
+                output_str += "\n```\n"
+        return output_str
+
+    def get_fpv_summary_results(self, result):
+        colalign = ("center", ) * len(self.header)
+        table = [self.header]
+        summary = []
+        fpv_summary = result.get("fpv_summary")
+        if fpv_summary is None:
+            results_str = "No fpv_summary found\n"
+            summary.append("N/A")
+        else:
+            table.append([
```

Comment:
Minor nit, but it's probably cleaner to just define table here: `table = [self.header, [self.name... ]]` (including some new lines!). If it's not possible to get the formatting to look good, at least move `table = [self.header]` into this branch of the if/else, since it's not used elsewhere.

Reply:
Thanks Rupert, yeah, I agree it is clearner that way.